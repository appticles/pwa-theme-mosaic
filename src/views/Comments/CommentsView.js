import React from 'react';
import { Header, Comment, Icon, Container, Button, Modal } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import _ from 'lodash';
import PropTypes from 'prop-types';

import { addComment } from './actions';
import { commentPropType } from './reducers';
import CommentForm from '../../components/Form/index';
import './style.css';

const CommentsView = (props) => {
  let path = {};
  if (_.isNil(props.match.params.categorySlug) || _.isNil(props.match.params.categoryId)) {
    path = '/post/' + props.match.params.postSlug + '/' + props.match.params.postId;
  } else {
    path =
      '/category/' +
      props.match.params.categorySlug +
      '/' +
      props.match.params.categoryId +
      '/post/' +
      props.match.params.postSlug +
      '/' +
      props.match.params.postId;
  }

  const submit = (values) => {
    const params = {
      author_name: values.name,
      email: values.email,
      content: {
        rendered: values.content,
      },
      date: new Date(),
      post: props.match.params.postId,
    };
    addComment(params);
    console.log(values);
  };
  console.log(props);
  return (
    <Comment.Group>
      <Header as="h3" block>
        Comments
        <Link to={path}>
          <Icon name="close" />
        </Link>
      </Header>
      {props.comments.length > 0 ? (
        props.comments.map(comment => (
          <Comment key={comment.id + comment.author}>
            <Comment.Avatar src={comment.author_avatar_urls[24]} />
            <Comment.Content>
              <Comment.Author as="a">{comment.author_name}</Comment.Author>
              <Comment.Metadata>
                <div>{comment.date}</div>
              </Comment.Metadata>
              <Comment.Text dangerouslySetInnerHTML={{ __html: comment.content.rendered }} />
              <Comment.Actions>
                <Comment.Action>Reply</Comment.Action>
              </Comment.Actions>
            </Comment.Content>
          </Comment>
        ))
      ) : (
        <Container className="no-comments">There are no comments for this article. Be the first one to comment about this!</Container>
      )}
      {props.match.params.comment_status === 'open' ? (
        <Modal trigger={<Button>Leave a comment</Button>} closeIcon>
          <Header icon="commenting" content="Leave a comment" />
          <Modal.Content>
            <CommentForm onSubmit={submit} />
          </Modal.Content>
        </Modal>
      ) : null}
    </Comment.Group>
  );
};

CommentsView.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      postSlug: PropTypes.string.isRequired,
      postId: PropTypes.string.isRequired,
      categorySlug: PropTypes.string,
      categoryId: PropTypes.string,
      comment_status: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  comments: PropTypes.arrayOf(commentPropType).isRequired,
};

export default CommentsView;
