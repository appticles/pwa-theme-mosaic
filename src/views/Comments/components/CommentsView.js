import React, { Component } from 'react';
import { Header, Comment, Icon, Container, Button, Modal, Loader } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import ReactMoment from 'react-moment';
import Moment from 'moment';
import config from 'react-global-configuration';
import { addComment } from '../actions';
import { commentPropType } from '../reducers';
import CommentForm from '../../../components/Form/index';

import './style.css';

class CommentsView extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
    };

    this.submit = this.submit.bind(this);
    this.showModal = this.showModal.bind(this);
  }

  submit(values) {
    const params = {
      author_name: values.name,
      author_email: values.email,
      content: values.content,
      date: new Moment().format(),
      post: Number(this.props.postId),
    };
    addComment(params);
  }

  showModal() {
    this.setState({ isOpen: !this.state.isOpen });
  }

  render() {
    if (this.props.loading) {
      return (
        <Comment.Group className="post-comments">
          <Loader active />
        </Comment.Group>
      );
    }

    return (
      <Comment.Group className="post-comments">
        {this.props.comments.length > 0 ? (
          this.props.comments.map(comment => (
            <Comment key={String(comment.id) + String(comment.content.rendered.length)}>
              <Comment.Avatar src={comment.author_avatar_urls[24]} />
              <Comment.Content>
                <Comment.Author as="a" href={comment.author_url}>
                  {comment.author_name}
                </Comment.Author>
                <Comment.Metadata>
                  <div>
                    <ReactMoment format="MMMM DD, YYYY">{comment.date}</ReactMoment>
                  </div>
                </Comment.Metadata>
                <Comment.Text dangerouslySetInnerHTML={{ __html: comment.content.rendered }} />
                <Comment.Actions>
                  <Comment.Action onClick={this.showModal}>
                    {this.props.texts.FORMS && this.props.texts.FORMS.REPLY}
                  </Comment.Action>
                </Comment.Actions>
              </Comment.Content>
            </Comment>
          ))
        ) : (
          <Container className="no-comments">
            {this.props.commentStatus === 'open'
              ? this.props.texts.TEXTS && this.props.texts.TEXTS.NO_COMMENTS
              : this.props.texts.TEXTS && this.props.texts.TEXTS.NO_COMMENTS_SHORT}
          </Container>
        )}

        {config.get('commentsEnabled') === 1 && this.props.commentStatus === 'open' ? (
          <Button onClick={this.showModal}>
            {this.props.texts.TEXTS && this.props.texts.TEXTS.LEAVE_COMMENTS}
          </Button>
        ) : null}

        {this.state.isOpen === true ? (
          <Modal open className="comments-modal">
            <Icon link name="close" onClick={this.showModal} />
            <Header
              icon="commenting"
              content={this.props.texts.TEXTS && this.props.texts.TEXTS.LEAVE_COMMENTS}
            />
            <Modal.Content>
              <CommentForm onSubmit={this.submit} texts={this.props.texts} />
            </Modal.Content>
          </Modal>
        ) : null}
      </Comment.Group>
    );
  }
}

CommentsView.propTypes = {
  postId: PropTypes.string.isRequired,
  commentStatus: PropTypes.oneOf(['open', 'closed', 'disabled']).isRequired,
  comments: PropTypes.arrayOf(commentPropType).isRequired,
  loading: PropTypes.number.isRequired,
  texts: PropTypes.shape({
    FORMS: PropTypes.shape({
      REPLY: PropTypes.string,
    }),
    TEXTS: PropTypes.shape({
      NO_COMMENTS: PropTypes.string,
      NO_COMMENTS_SHORT: PropTypes.string,
      LEAVE_COMMENTS: PropTypes.string,
    }),
  }),
};

CommentsView.defaultProps = {
  texts: {
    FORMS: {
      REPLY: 'Reply',
    },
    TEXTS: {
      NO_COMMENTS: 'No comments',
      NO_COMMENTS_SHORT: 'No comments',
      LEAVE_COMMENTS: 'Leave comment',
    },
  },
};

export default CommentsView;
