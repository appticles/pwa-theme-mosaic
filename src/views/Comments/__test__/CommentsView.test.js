import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import CommentsView from '../components/CommentsView';

describe('Comments view', () => {
  it('renders no comments, allows adding new comments', () => {
    const comments = [];
    const match = {
      params: {
        postSlug: 'post-slug',
        postId: '43',
        categorySlug: 'category-slug',
        categoryId: '54',
        comment_status: 'open',
      },
    };

    const output = shallow(<CommentsView comments={comments} match={match} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
  it('renders no comments, does not allow adding new comments', () => {
    const comments = [];
    const match = {
      params: {
        postSlug: 'post-slug',
        postId: '43',
        categorySlug: 'category-slug',
        categoryId: '54',
        comment_status: 'close',
      },
    };

    const output = shallow(<CommentsView comments={comments} match={match} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
  it('renders comments, allows adding new comments', () => {
    const comments = [
      {
        id: 67,
        post: 54,
        author_name: 'Comment Author',
        content: {
          rendered: 'Comment content',
        },
        date: 'd date',
        author_avatar_urls: {
          24: 'some-rurl',
        },
      },
    ];
    const match = {
      params: {
        postSlug: 'post-slug',
        postId: '43',
        categorySlug: 'category-slug',
        categoryId: '54',
        comment_status: 'open',
      },
    };
    const output = shallow(<CommentsView comments={comments} match={match} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
  it('renders comments, does not allow adding new comments', () => {
    const comments = [
      {
        id: 67,
        post: 54,
        author_name: 'Comment Author',
        content: {
          rendered: 'Comment content',
        },
        date: 'd date',
        author_avatar_urls: {
          24: 'some-rurl',
        },
      },
    ];
    const match = {
      params: {
        postSlug: 'post-slug',
        postId: '43',
        categorySlug: 'category-slug',
        categoryId: '54',
        comment_status: 'close',
      },
    };
    const output = shallow(<CommentsView comments={comments} match={match} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});
