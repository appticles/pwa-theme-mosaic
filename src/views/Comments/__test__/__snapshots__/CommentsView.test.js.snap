// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments view renders comments, allows adding new comments 1`] = `
<CommentGroup>
  <Comment
    key="6715"
  >
    <CommentAvatar
      src="some-rurl"
    />
    <CommentContent>
      <CommentAuthor
        as="a"
      >
        Comment Author
      </CommentAuthor>
      <CommentMetadata>
        <div>
          <t
            ago={false}
            calendar={false}
            element="time"
            filter={[Function]}
            format="MMMM DD, YYYY"
            fromNow={false}
            interval={60000}
            onChange={[Function]}
            toNow={false}
            unix={false}
            utc={false}
          >
            d date
          </t>
        </div>
      </CommentMetadata>
      <CommentText
        dangerouslySetInnerHTML={
          Object {
            "__html": "Comment content",
          }
        }
      />
      <CommentActions>
        <CommentAction
          as="a"
          onClick={[Function]}
        >
          Reply
        </CommentAction>
      </CommentActions>
    </CommentContent>
  </Comment>
  <Button
    as="button"
    onClick={[Function]}
  >
    Leave comment
  </Button>
</CommentGroup>
`;

exports[`Comments view renders comments, does not allow adding new comments 1`] = `
<CommentGroup>
  <Comment
    key="6715"
  >
    <CommentAvatar
      src="some-rurl"
    />
    <CommentContent>
      <CommentAuthor
        as="a"
      >
        Comment Author
      </CommentAuthor>
      <CommentMetadata>
        <div>
          <t
            ago={false}
            calendar={false}
            element="time"
            filter={[Function]}
            format="MMMM DD, YYYY"
            fromNow={false}
            interval={60000}
            onChange={[Function]}
            toNow={false}
            unix={false}
            utc={false}
          >
            d date
          </t>
        </div>
      </CommentMetadata>
      <CommentText
        dangerouslySetInnerHTML={
          Object {
            "__html": "Comment content",
          }
        }
      />
      <CommentActions>
        <CommentAction
          as="a"
          onClick={[Function]}
        >
          Reply
        </CommentAction>
      </CommentActions>
    </CommentContent>
  </Comment>
</CommentGroup>
`;

exports[`Comments view renders no comments, allows adding new comments 1`] = `
<CommentGroup>
  <Container
    className="no-comments"
  >
    No comments
  </Container>
  <Button
    as="button"
    onClick={[Function]}
  >
    Leave comment
  </Button>
</CommentGroup>
`;

exports[`Comments view renders no comments, does not allow adding new comments 1`] = `
<CommentGroup>
  <Container
    className="no-comments"
  >
    No comments
  </Container>
</CommentGroup>
`;
